src/main.cpp

```diff
@@ -179,6 +189,165 @@ BOOL CtrlHandler(DWORD fdwCtrlType)
+// serviceWorkerThread: Keep the service alive
+
+DWORD WINAPI serviceWorkerThread(LPVOID lpParam)
+{
+    UNREFERENCED_PARAMETER(lpParam);
+
+    //Do the normal startup
+    universalMain();
+
+    return ERROR_SUCCESS;
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceSetState: Set the internal status of the service
+
+void serviceSetState(DWORD acceptedControls, DWORD newState, DWORD exitCode)
+{
+    SERVICE_STATUS serviceStatus;
+    ZeroMemory(&serviceStatus, sizeof(SERVICE_STATUS));
+    serviceStatus.dwCheckPoint = 0;
+    serviceStatus.dwControlsAccepted = acceptedControls;
+    serviceStatus.dwCurrentState = newState;
+    serviceStatus.dwServiceSpecificExitCode = 0;
+    serviceStatus.dwServiceType = SERVICE_WIN32;
+    serviceStatus.dwWaitHint = 0;
+    serviceStatus.dwWin32ExitCode = exitCode;
+
+    if (SetServiceStatus(g_StatusHandle, &serviceStatus) == FALSE)
+    {
+        LOGERROR("SetServiceStatus() failed\n");
+    }
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceCtrlHandler: Handle stop events from the Service Control Manager
+
+void WINAPI serviceCtrlHandler(DWORD CtrlCode)
+{
+    switch (CtrlCode)
+    {
+    case SERVICE_CONTROL_STOP:
+        Root.SetStopping(true);
+        serviceSetState(0, SERVICE_STOP_PENDING, 0);
+        break;
+
+    default:
+        break;
+    }
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceMain: Startup logic for running as a service
+
+void WINAPI serviceMain(DWORD argc, TCHAR *argv[])
+{
+    Sleep(10000);
+
+    char* applicationFilename = (char*)calloc(MAX_PATH, sizeof(char));
+    char* applicationDirectory = (char*)calloc(MAX_PATH, sizeof(char));
+
+    GetModuleFileName(NULL, applicationFilename, MAX_PATH); //This binaries fill path.
+
+    //GetModuleFileName() returns the path and filename. Strip off the filename.
+    strncpy(applicationDirectory, applicationFilename, strrchr(applicationFilename, '\\') - applicationFilename);
+    applicationDirectory[strlen(applicationDirectory)] = '\0'; //Make sure new path is null terminated
+
+    //Services are run by the SCM, and inherit its working directory - usually System32.
+    // Set the working directory to the same location as the binary.
+    SetCurrentDirectory(applicationDirectory); 
+
+    g_StatusHandle = RegisterServiceCtrlHandler(SERVICE_NAME, serviceCtrlHandler);
+
+    if (g_StatusHandle == NULL)
+    {
+        LOGERROR("RegisterServiceCtrlHandler() failed\n");
+        serviceSetState(0, SERVICE_STOPPED, GetLastError());
+        return;
+    }
+
+    serviceSetState(SERVICE_ACCEPT_STOP, SERVICE_RUNNING, 0);
+
+    g_ServiceThread = CreateThread(NULL, 0, serviceWorkerThread, NULL, 0, NULL);
+    if (g_ServiceThread == NULL)
+    {
+        LOGERROR("CreateThread() failed\n");
```

Comment:
This won't log anywhere because the logger listeners aren't attached yet. I'm not sure where you can log to though.